src/ClientMessage.ts

```diff
@@ -15,248 +15,214 @@
- |        Data Offset          |                                 |
- +-----------------------------+                                 |
- |                      Message Payload Data                    ...
- |                                                              ...
- */
-
 import {Buffer} from 'safe-buffer';
 import * as Long from 'long';
 import {BitsUtil} from './BitsUtil';
-import {Data} from './serialization/Data';
-import {HeapData} from './serialization/HeapData';
-
-class ClientMessage {
+import {ClientConnection} from './network/ClientConnection';
 
-    private buffer: Buffer;
-    private cursor: number = BitsUtil.HEADER_SIZE;
-    private isRetryable: boolean;
+const MESSAGE_TYPE_OFFSET = 0;
+const CORRELATION_ID_OFFSET = MESSAGE_TYPE_OFFSET + BitsUtil.INT_SIZE_IN_BYTES;
+export const RESPONSE_BACKUP_ACKS_OFFSET = CORRELATION_ID_OFFSET + BitsUtil.LONG_SIZE_IN_BYTES;
+export const PARTITION_ID_OFFSET = CORRELATION_ID_OFFSET + BitsUtil.LONG_SIZE_IN_BYTES;
+const FRAGMENTATION_ID_OFFSET = 0;
 
-    constructor(buffer: Buffer) {
-        this.buffer = buffer;
-    }
-
-    public static newClientMessage(payloadSize: number): ClientMessage {
-        const totalSize = BitsUtil.HEADER_SIZE + payloadSize;
-        const buffer = Buffer.allocUnsafe(totalSize);
-        const message = new ClientMessage(buffer);
-        message.setDataOffset(BitsUtil.HEADER_SIZE);
-        message.setVersion(BitsUtil.VERSION);
-        message.setFrameLength(totalSize);
-        message.setFlags(0xc0);
-        message.setPartitionId(-1);
-        return message;
-    }
+const DEFAULT_FLAGS = 0;
+const BEGIN_FRAGMENT_FLAG = 1 << 15;
+const END_FRAGMENT_FLAG = 1 << 14;
+export const UNFRAGMENTED_MESSAGE = BEGIN_FRAGMENT_FLAG | END_FRAGMENT_FLAG;
+export const IS_FINAL_FLAG = 1 << 13;
+const BEGIN_DATA_STRUCTURE_FLAG = 1 << 12;
+const END_DATA_STRUCTURE_FLAG = 1 << 11;
+const IS_NULL_FLAG = 1 << 10;
+export const IS_EVENT_FLAG = 1 << 9;
+const BACKUP_AWARE_FLAG = 1 << 8;
+const BACKUP_EVENT_FLAG = 1 << 7;
 
-    clone(): ClientMessage {
-        const message = new ClientMessage(Buffer.from(this.buffer));
-        message.isRetryable = this.isRetryable;
-        return message;
-    }
+export const SIZE_OF_FRAME_LENGTH_AND_FLAGS = BitsUtil.INT_SIZE_IN_BYTES + BitsUtil.SHORT_SIZE_IN_BYTES;
 
-    getBuffer(): Buffer {
-        return this.buffer;
-    }
+export class Frame {
+    content: Buffer;
+    flags: number;
+    next: Frame;
 
-    getCorrelationId(): number {
-        const offset = BitsUtil.CORRELATION_ID_FIELD_OFFSET;
-        return this.readLongInternal(offset).toNumber();
+    constructor(content: Buffer, flags?: number) {
+        this.content = content;
+        this.flags = flags || DEFAULT_FLAGS;
     }
 
-    setCorrelationId(value: number): void {
-        this.writeLongInternal(value, BitsUtil.CORRELATION_ID_FIELD_OFFSET);
+    static createInitialFrame(size: number, flags?: number): Frame {
```

Comment:
nit: it seems that this method is always called with `UNFRAGMENTED_MESSAGE` as the second argument. If that's the case, we could use this as the default value for `flags` argument, when it's not specified. That could help us to keep all flags internal to this module.

Reply:
This method is also used in custom codecs which have `DEFAULT_FLAGS` as their flags. I added `UNFRAGMENTED_MESSAGE` as the default parameter as you suggested and used `undefined` as the flag for custom types so that they will have `DEFAULT_FLAGS`  (due to constructor of `Frame` class.). What do you think about it ? Maybe we can introduce another static factory method or expose `DEFAULT_FLAGS` so that we can use it on custom codecs ?

Reply:
> Maybe we can introduce another static factory method or expose `DEFAULT_FLAGS` so that we can use it on custom codecs ?

I'd vote for exposing `DEFAULT_FLAGS` in this case. Passing `undefined` as the second argument makes the code complicated.

Reply:
Exposed the `DEFAULT_FLAGS` and used in the custom codecs